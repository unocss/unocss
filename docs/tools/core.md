---
title: Core
description: The core engine of UnoCSS without any presets. It can be used as the engine of your own atomic CSS framework.
---

# Core

The core engine of UnoCSS without any presets: `@unocss/core`. It can be used as the engine of your own atomic CSS framework.

## Installation

::: code-group

```bash [pnpm]
pnpm add -D @unocss/core
```

```bash [yarn]
yarn add -D @unocss/core
```

```bash [npm]
npm install -D @unocss/core
```

```bash [bun]
bun add -D @unocss/core
```

:::

## Usage

```ts
import { createGenerator } from '@unocss/core'

const generator = await createGenerator(
  { /* user options */ },
  { /* default options */ }
)

const { css } = await generator.generate(code)
```

## License

- MIT License &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
