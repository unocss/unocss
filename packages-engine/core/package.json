{
  "name": "@unocss/core",
  "type": "module",
  "version": "66.1.0-beta.3",
  "description": "The instant on-demand Atomic CSS engine.",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://unocss.dev",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss",
    "directory": "packages-engine/core"
  },
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "keywords": [
    "unocss",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "dist/index.mjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub",
    "test:attw": "attw --pack --config-path ../../.attw-esm-only.json"
  },
  "devDependencies": {
    "magic-string": "catalog:",
    "unconfig": "catalog:"
  }
}
