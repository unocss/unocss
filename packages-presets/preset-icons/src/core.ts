import type { IconifyJSON } from '@iconify/types'
import type { IconifyLoaderOptions, UniversalIconLoader } from '@iconify/utils'
import type { CSSObject } from '@unocss/core'
import type { IconsOptions } from './types'
import { loadIcon } from '@iconify/utils/lib/loader/loader'
import { searchForIcon } from '@iconify/utils/lib/loader/modern'
import { encodeSvgForCss } from '@iconify/utils/lib/svg/encode-svg-for-css'
import { definePreset, warnOnce } from '@unocss/core'
import icons from './collections'

const COLLECTION_NAME_PARTS_MAX = 3

export { IconsOptions }
export { icons }

/**
 * API for preset-icons
 */
export interface IconsAPI {
  encodeSvgForCss: typeof encodeSvgForCss
  parseIconWithLoader: typeof parseIconWithLoader
  /**
   * This API only available for preset-icons created on Node.js environment
   */
  createNodeLoader?: () => Promise<UniversalIconLoader | undefined>
}

export function createPresetIcons(lookupIconLoader: (options: IconsOptions) => Promise<UniversalIconLoader>) {
  return definePreset((options: IconsOptions = {}) => {
    const {
      scale = 1,
      mode = 'auto',
      prefix = 'i-',
      warn = false,
      iconifyCollectionsNames,
      collections: customCollections,
      extraProperties = {},
      customizations = {},
      autoInstall = false,
      collectionsNodeResolvePath,
      layer = 'icons',
      unit,
      processor,
    } = options

    const flags = getEnvFlags()

    const loaderOptions: IconifyLoaderOptions = {
      addXmlNs: true,
      scale,
      customCollections,
      autoInstall,
      cwd: collectionsNodeResolvePath,
      // avoid warn from @iconify/loader: we'll warn below if not found
      warn: undefined,
      customizations: {
        ...customizations,
        additionalProps: { ...extraProperties },
        trimCustomSvg: true,
        async iconCustomizer(collection, icon, props) {
          await customizations.iconCustomizer?.(collection, icon, props)
          if (unit) {
            if (!props.width)
              props.width = `${scale}${unit}`
            if (!props.height)
              props.height = `${scale}${unit}`
          }
        },
      },
    }

    let iconLoader: UniversalIconLoader

    return {
      name: '@unocss/preset-icons',
      enforce: 'pre',
      options,
      layers: { icons: -30 },
      api: <IconsAPI>{
        encodeSvgForCss,
        parseIconWithLoader,
      },
      rules: [[
        /^([\w:-]+)(?:\?(mask|bg|auto))?$/,
        async (matcher) => {
          let [full, body, _mode = mode] = matcher as [string, string, IconsOptions['mode']]

          iconLoader = iconLoader || await lookupIconLoader(options)

          const usedProps = {}
          const parsed = await parseIconWithLoader(
            body,
            iconLoader,
            { ...loaderOptions, usedProps },
            iconifyCollectionsNames,
          )

          if (!parsed) {
            if (warn && !flags.isESLint)
              warnOnce(`failed to load icon "${full}"`)
            return
          }

          let cssObject: CSSObject
          const url = `url("data:image/svg+xml;utf8,${encodeSvgForCss(parsed.svg)}")`

          if (_mode === 'auto')
            _mode = parsed.svg.includes('currentColor') ? 'mask' : 'bg'

          if (_mode === 'mask') {
            // Thanks to https://codepen.io/noahblon/post/coloring-svgs-in-css-background-images
            cssObject = {
              '--un-icon': url,
              '-webkit-mask': 'var(--un-icon) no-repeat',
              'mask': 'var(--un-icon) no-repeat',
              '-webkit-mask-size': '100% 100%',
              'mask-size': '100% 100%',
              'background-color': 'currentColor',
              // for Safari https://github.com/elk-zone/elk/pull/264
              'color': 'inherit',
              ...usedProps,
            }
          }
          else {
            cssObject = {
              'background': `${url} no-repeat`,
              'background-size': '100% 100%',
              'background-color': 'transparent',
              ...usedProps,
            }
          }

          processor?.(cssObject, {
            ...parsed,
            icon: parsed.name,
            mode: _mode,
          })

          return cssObject
        },
        { layer, prefix },
      ]],
    }
  })
}

export function combineLoaders(loaders: UniversalIconLoader[]) {
  return (async (...args) => {
    for (const loader of loaders) {
      if (!loader)
        continue
      const result = await loader(...args)
      if (result)
        return result
    }
  }) as UniversalIconLoader
}

export function createCDNFetchLoader(
  fetcher: (url: string) => Promise<any>,
  cdnBase: string,
  cacheMap: Map<string, IconifyJSON | Promise<IconifyJSON>> = new Map(),
): UniversalIconLoader {
  function fetchCollection(name: string) {
    if (!icons.includes(name))
      return undefined
    if (!cacheMap.has(name))
      cacheMap.set(name, fetcher(`${cdnBase}@iconify-json/${name}/icons.json`))
    return cacheMap.get(name)!
  }

  return async (collection, icon, options) => {
    let result = await loadIcon(collection, icon, options)
    if (result)
      return result

    const iconSet = await fetchCollection(collection)
    if (iconSet) {
      // possible icon names
      const ids = [
        icon,
        icon.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase(),
        icon.replace(/([a-z])(\d+)/g, '$1-$2'),
      ]
      result = await searchForIcon(iconSet, collection, ids, options)
    }

    return result
  }
}

export function getEnvFlags() {
  // eslint-disable-next-line node/prefer-global/process
  const isNode = typeof process !== 'undefined' && process.stdout && !process.versions.deno
  // eslint-disable-next-line node/prefer-global/process
  const isVSCode = isNode && !!process.env.VSCODE_CWD
  // eslint-disable-next-line node/prefer-global/process
  const isESLint = isNode && !!process.env.ESLINT

  return {
    isNode,
    isVSCode,
    isESLint,
  }
}

export async function parseIconWithLoader(
  body: string,
  loader: UniversalIconLoader,
  options: IconifyLoaderOptions = {},
  safeCollectionsNames: string[] = [],
) {
  let collection = ''
  let name = ''
  let svg: string | undefined

  const allCollections = new Set<string>([
    ...icons,
    ...safeCollectionsNames,
    ...Object.keys(options.customCollections || {}),
  ])

  if (body.includes(':')) {
    [collection, name] = body.split(':')

    if (!allCollections.has(collection))
      return

    svg = await loader(collection, name, options)
  }
  else {
    const parts = body.split(/-/g)
    for (let i = COLLECTION_NAME_PARTS_MAX; i >= 1; i--) {
      collection = parts.slice(0, i).join('-')

      if (!allCollections.has(collection))
        continue

      name = parts.slice(i).join('-')
      svg = await loader(collection, name, options)

      if (svg)
        break
    }
  }

  if (!svg) {
    return
  }

  return {
    collection,
    name,
    svg,
  }
}
