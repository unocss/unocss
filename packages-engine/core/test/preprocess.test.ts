import { createGenerator } from '@unocss/core'
import presetIcons from '@unocss/preset-icons'
import presetWind3 from '@unocss/preset-wind3'
import { describe, expect, it } from 'vitest'

describe('preprocess', () => {
  it('prefix', async () => {
    const positive = [
      'uno:pl-10px',
      'uno:hover:p-4',
      'uno:sm:p-6',
      'uno:!p-5px',
      'uno:btn',
      'uno:btn1',
      'uno:dark:i-carbon-moon',
    ]

    const negative = [
      'pl-10px',
      'hover:p-4',
      '!p-5px',
      'btn',
      'btn1',
      'i-carbon-moon',
    ]

    const uno = await createGenerator({
      preprocess: m => m.startsWith('uno:') ? m.substr(4) : '',
      presets: [
        presetWind3(),
        presetIcons(),
      ],
      shortcuts: {
        btn: 'mr-10',
        btn1: 'ml-10 btn',
      },
    })
    const { css, matched } = await uno.generate(new Set([...positive, ...negative]), { preflights: false })
    expect(matched).eql(new Set(positive))
    expect(css).toMatchSnapshot()
  })

  it('tailwind prefix', async () => {
    const positive = [
      'uno:pl-10px',
      'hover:uno:p-4',
      'sm:uno:p-6',
      '!uno:p-5px',
      'uno-btn',
      'uno:btn1',
    ]

    const negative = [
      'pl-10px',
      'hover:p-4',
      '!p-5px',
      'btn',
      'btn1',
    ]

    const prefixRE = /uno[:-]/
    const uno = await createGenerator({
      preprocess: m => prefixRE.test(m) ? m.replace(prefixRE, '') : '',
      presets: [
        presetWind3(),
        presetIcons(),
      ],
      shortcuts: {
        btn: 'mr-10',
        btn1: 'ml-10 btn',
      },
    })
    const { css, matched } = await uno.generate(new Set([...positive, ...negative]), { preflights: false })
    expect(matched).eql(new Set(positive))
    expect(css).toMatchSnapshot()
  })
})
