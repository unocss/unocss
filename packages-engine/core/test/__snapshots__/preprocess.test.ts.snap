// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preprocess > prefix 1`] = `
"/* layer: icons */
.dark .uno\\:dark\\:i-carbon-moon{--un-icon:url("data:image/svg+xml;utf8,%3Csvg viewBox='0 0 32 32' width='1em' height='1em' xmlns='http://www.w3.org/2000/svg' %3E%3Cpath fill='currentColor' d='M13.503 5.414a15.076 15.076 0 0 0 11.593 18.194a11.1 11.1 0 0 1-7.975 3.39c-.138 0-.278.005-.418 0a11.094 11.094 0 0 1-3.2-21.584M14.98 3a1 1 0 0 0-.175.016a13.096 13.096 0 0 0 1.825 25.981c.164.006.328 0 .49 0a13.07 13.07 0 0 0 10.703-5.555a1.01 1.01 0 0 0-.783-1.565A13.08 13.08 0 0 1 15.89 4.38A1.015 1.015 0 0 0 14.98 3'/%3E%3C/svg%3E");-webkit-mask:var(--un-icon) no-repeat;mask:var(--un-icon) no-repeat;-webkit-mask-size:100% 100%;mask-size:100% 100%;background-color:currentColor;color:inherit;width:1em;height:1em;}
/* layer: shortcuts */
.uno\\:btn{margin-right:2.5rem;}
.uno\\:btn1{margin-left:2.5rem;margin-right:2.5rem;}
/* layer: default */
.uno\\:\\!p-5px{padding:5px !important;}
.uno\\:hover\\:p-4:hover{padding:1rem;}
.uno\\:pl-10px{padding-left:10px;}
.uno\\:sm\\:p-6{@media (min-width: 640px){
padding:1.5rem;
}}"
`;

exports[`preprocess > tailwind prefix 1`] = `
"/* layer: shortcuts */
.uno-btn{margin-right:2.5rem;}
.uno\\:btn1{margin-left:2.5rem;margin-right:2.5rem;}
/* layer: default */
.\\!uno\\:p-5px{padding:5px !important;}
.hover\\:uno\\:p-4:hover{padding:1rem;}
.uno\\:pl-10px{padding-left:10px;}
.sm\\:uno\\:p-6{@media (min-width: 640px){
padding:1.5rem;
}}"
`;
